### PR TITLE
DDBTEAM-421: Added support for reindex search with clean-up

### PR DESCRIPTION
**Pull request by cableman**
https://jira.itkdev.dk/browse/DDBTEAM-421

### CODE REVIEW

File: src/Command/Search/SearchReindexCommand.php [hunk 1]
```diff
@@ -37,7 +43,8 @@ public function __construct(EntityManagerInterface $entityManager, ProducerInter
     protected function configure()
     {
         $this->setDescription('Reindex search table')
-            ->addArgument('vendorid', InputArgument::OPTIONAL, 'Limit the re-index to vendor with this id number');
+            ->addArgument('vendorid', InputArgument::OPTIONAL, 'Limit the re-index to vendor with this id number')
+            ->addOption('clean-up', null, InputOption::VALUE_NONE, 'Remove all rows from the search table related to an given source before insert');
```
Comment (tuj) [inline src/Command/Search/SearchReindexCommand.php]: an => a

File: src/Queue/SearchProcessor.php [hunk 2]
```diff
@@ -55,8 +58,25 @@ public function __construct(EntityManagerInterface $entityManager, EventDispatch
     public function process(Message $message, Context $session)
     {
         $jsonDecoder = new JsonDecoder(true);
+        /* @var ProcessMessage $processMessage*/
```
Comment (tuj) [inline src/Queue/SearchProcessor.php]: Missing space before */

File: src/Queue/SearchProcessor.php [hunk 3]
```diff
@@ -55,8 +58,25 @@ public function __construct(EntityManagerInterface $entityManager, EventDispatch
...
         $processMessage = $jsonDecoder->decode($message->getBody(), ProcessMessage::class);
 
+        // Clean up: find all search that links back to a give source and remove them before sending new index event.
+        // This is done even if the search below is an zero-hit.
```
Comment (tuj) [inline src/Queue/SearchProcessor.php]: an => a

File: src/Queue/SearchProcessor.php [hunk 4]
```diff
@@ -55,8 +58,25 @@ public function __construct(EntityManagerInterface $entityManager, EventDispatch
...
+        if (VendorState::DELETE_AND_UPDATE === $processMessage->getOperation()) {
+            $sourceRepos = $this->em->getRepository(Source::class);
+            /* @var Source $source */
+            $source = $sourceRepos->findOneBy([
+                'matchId' => $processMessage->getIdentifier(),
+                'matchType' => $processMessage->getIdentifierType(),
+            ]);
+            $searches = $source->getSearches();
```
Comment (tuj) [inline src/Queue/SearchProcessor.php]: What if source is null?

File: src/Service/OpenPlatform/SearchService.php [hunk 5]
```diff
@@ -259,7 +259,7 @@ private function recursiveSearch(string $token, string $identifier, string $type
         switch ($type) {
             case IdentifierType::PID:
                 // If this is a search after a pid simply search for it and not in the search index.
-                $query = $identifier;
+                $query = 'rec.id='.$identifier;
```
Comment (tuj) [inline src/Service/OpenPlatform/SearchService.php]: Is this related to reindex?

File: src/Utils/Types/VendorState.php [hunk 6]
```diff
@@ -7,4 +7,5 @@ class VendorState
     const INSERT = 'insert';
     const UPDATE = 'update';
     const DELETE = 'delete';
+    const DELETE_AND_UPDATE = 'DeleteAndUpdate';
```
Comment (tuj) [inline src/Utils/Types/VendorState.php]: Why is this camel case, when the others are lowercase?

File: src/Queue/SearchProcessor.php (hunk 4 above)
Comment (cableman) [reply src/Queue/SearchProcessor.php]: It's never null at this point... if it is we have an corrupted database

File: src/Service/OpenPlatform/SearchService.php (hunk 5 above)
Comment (cableman) [reply src/Service/OpenPlatform/SearchService.php]: Yes. It's a change that requires us to re-index

File: src/Utils/Types/VendorState.php (hunk 6 above)
Comment (cableman) [reply src/Utils/Types/VendorState.php]: Changed first word to lower case.

File: src/Queue/SearchProcessor.php (hunk 3 above)
Comment (cableman) [reply src/Queue/SearchProcessor.php]: Fixed

File: src/Queue/SearchProcessor.php (hunk 2 above)
Comment (cableman) [reply src/Queue/SearchProcessor.php]: Fixed

File: src/Command/Search/SearchReindexCommand.php (hunk 1 above)
Comment (cableman) [reply src/Command/Search/SearchReindexCommand.php]: Fixed

File: src/Queue/SearchProcessor.php (hunk 4 above)
Comment (cableman) [reply src/Queue/SearchProcessor.php]: Fixed
